### ISSUE TITLE
Enums with duplicate entries are incorrectly deduplicated

### ISSUE THREAD

**Issue by ouvreboite** [domaindrivendev/Swashbuckle.AspNetCore#2377]
**Version**: 6.2.3
**Using NewtonSoft** (and settings.Converters.Add(new StringEnumConverter())

Enums with multiple values for the same ordinal (for example System.Net.HttpStatusCode which have **Moved** = **MovedPermanently** = 301) are deduplicated.

In older versions (~4) both Moved and MovedPermanently would be listed
```
"enum": [ 
   ...,
   "Moved",
   "MovedPermanently",
   ...
]
```

In this version (6.2.3), only one appears:
```
"enum": [ 
   ...,
   "Moved",
   ...
]
```

From what I see, the problem may come from SchemaGenerator.CreatePrimitiveSchema
```
if (dataContract.UnderlyingType.IsEnum)
{
    schema.Enum = dataContract.UnderlyingType.GetEnumValues()
        .Cast<object>()
        .Distinct()
        .Select(value => dataContract.JsonConverter(value))
        .Select(valueAsJson => OpenApiAnyFactory.CreateFromJson(valueAsJson))
        .ToList();
}
```

UnderlyingType.GetEnumValues() returns correctly all the enum values, but **Moved** and **MovedPermanently** having the same ordinal they are the "same" object and so one of them is filtered by the .Distinct()

**Comment by ouvreboite** [comment]
One short-term fix is to create a custom SchemaFilter that uses enum.GetEnumNames().

```
swaggerGenOptions.SchemaFilter<DuplicateEnumFilter>();

...

public class DuplicateEnumFilter : ISchemaFilter
{
    public void Apply(OpenApiSchema schema, SchemaFilterContext context)
    {
        if (context.Type.IsEnum)
        {
            schema.Enum = context.Type.GetEnumNames()
                .Select(enumName => (IOpenApiAny) new OpenApiString(enumName))
                .ToList();
        }
    }
}
```

**Comment by domaindrivendev** [comment]
I believe this is fixed in the latest version - please try that.

**Comment by ouvreboite** [comment]
I created a small sample project (condensed [Gist](https://gist.github.com/ouvreboite/ebbaa47825ef0ec3510098b69db1b5c7)) and tested both 6.2.3 and 6.3.0. They both return only "MovedPermanently".

The [fix](https://github.com/domaindrivendev/Swashbuckle.AspNetCore/commit/fdf8d9ad1794c608a16047de940ae2ad8bfc5066) that has been included in 6.3.0 simply applies the distinct earlier. But as dataContract.UnderlyingType.GetEnumValues() return the same "object" for enums that share an ordinal, it still discards duplicates.

**Comment by domaindrivendev** [comment]
Actually, I think this may be related to an issue (of sorts) with the STJ serializer and not Swashbuckle. Using your own example, if you update your action method as follows and then "try it out" from the swagger-ui (or curl to illustrate how this is unrelated to Swashbuckle), you'll notice (suprisingly) that the value `MovedPermanently` is returned in the JSON instead of `Moved`.

```csharp
[HttpGet("{id}")]
public async Task<ActionResult<SchemaWithEnum>> Get(long id)
{
    return new SchemaWithEnum()
    {
        Id = id,
        StatusCode = HttpStatusCode.Moved
    };
}  
```

```json
GET /api/enum/123
{
    "id": 123,
    "statusCode": "MovedPermanently" // WTF!!!
}
```

This is because at runtime there's just one value and STJ doesn't/can't really know which of the variations it is, so it just picks the last. And, it's this behavior that's ultimately causing the bug you're reporting. If you look closely at the Swashbuckle code, it maps all of the enum values to their corresponding "serialized form", determined by simply invoking the serializer, and then it calls `Distinct()`. But, because of the STJ behavior I described, the list that's passed to `Distinct()` won't contain both "Moved" and "MovedPermanently", it will contain "MovedPermanently" twice!

Even if we removed the deduping entirely, the Swagger would never list the "Moved" value - it would just show "MovedPermanently" twice!

Long story short - I don't think you should use enums with duplicate entries in your API contract because serializer behavior is unpredictable or at the very least unintuitive. On that note, I used your ecample but this time configuring for Newtonsoft instead of STJ and noticed it has the same issue but instead of choosing the last duplicate it chooses the first. So, you end up always getting "Moved" instead of "MovedPermanently".

**Comment by ouvreboite** [comment]
Enum's with duplicate entries are indeed code smell. But they can happen.

The problem in my case is that the behavior in Swashbuckle 4 was to expose both Moved and MovedPermanently (or other duplicates, HttpStatusCode is a convenient example). As we use Swagger/OpenApi specs to generate SDK for our APIs, it means migrating to Swashbuckle 6 will remove some enum values from the generated code and will create potential breaking changes for our APIs' users.

The fact that the documented duplicate is not the one returned when calling the endpoint (as per your example), is also problematic, as it would lead to deserialization errors.

I'm aware that this "feature" is quite specific and seems complex to handle. Would you like me to try to address it (with a PR) or is this something you don't think is valuable to pursue?

**Comment by domaindrivendev** [comment]
Well ... the goal of Swashbuckle is to generate docs that refelect the _actual_ API behavior and for enum's the optimal way of doing this is to actually enumerate and serialize the available values - i.e. exactly what the code does today. I'm not interested in reverting to an implementation (i.e. SB 4) that enumerates the available values and then tries to "second guess" what the serializer will do.

I understand and appreciate your special case and the issue that the latest behavior is causing but I believe this is because of an API implementation problem as opposed to SB. For your case, I think you should just create an `ISchemaFilter` that replaces enum values using custom logic to solve your problem - i.e. you could probably just use `context.Type.GetEnumNames()`.
